### ISSUE TITLE
autodiff: Difficult to track semantics of variables, difficult to mix gradients

### ISSUE THREAD

**Issue by EricCousineau-TRI** [RobotLocomotion/drake#8927]
## Description

@TobiaMarcucci ran into an issue when working with `RigidBodyTree` where he had some expressions that depend on just `q` (e.g. `massMatrix`), and others that depend on `q` and `v` (e.g. `dynamicsBiasTerm`). In doing so, using the current forward-diff implementation in `AutoDiffXd`, he had to either (a) explicitly ensure that all autodiff variables were specified w.r.t. `[q, v]` or (b) fuse the outputs (w.r.t. `q` and w.r.t. `[q, v]`) by tracking indices.

## Reproduction

A simplified scalar case:

```c++
AutoDiffXd f(AutoDiffXd x) {
  return 2*x;
}

AutoDiffXd g(AutoDiffXd x, AutoDiffXd y) {
  return x + y;
}

// (a) Pre-allocate
AutoDiffXd x(1., VectorXd::Unit(2, 0));
AutoDiffXd y(2., VectorXd::Unit(2, 1));
auto expr = f(x) + g(x, y);

// (b) Fuse
AutoDiffXd x_x(1., VectorXd::Unit(1, 0));
auto f_x = f(x_x);
// ... Later point in the code
AutoDiffXd y_xy(1., VectorXd::Unit(2, 0));
AutoDiffXd y_xy(2., VectorXd::Unit(2, 1));
auto g_xy = g(x_xy, y_xy);
// Simple fusing:
AutoDiffXd f_xy(f_x.value(), (VectorXd(2) << f_x.derivatives(), 0).finished()));
auto expr = f_xy + g_xy;
```

For vector-valued cases, things like `initializeAutoDiff`, `autoDiffToGradientMatrix`, and `gradientMatrixToAutoDiff` would be used.

## Possible Solutions

1. Do workaround (a) or (b).
2. Suggestions from @jwnimmer-tri:
    a. If working within `MathematicalProgram`, leave the variable abstraction to upstream calls.
    b. If working with rigid body dynamics, consider using the systems framework (per the prior discussions, would require making a relaxed version of `Linearize`) and/or the `MultibodyTree` API once it is supported in Python.
3. For general math with AutoDiff, consider implementing a new scalar type, `AutoDiffSemantic`, which would record the values.
    a. One option is to implicitly concatenate values. However, this may cause silent bugs (e.g. computing `f(x, y)`, but only w.r.t. `x`, then later fusing that with values from a gradient from both `[x, y]`).
    b. Track the semantics, add explicit helpers to fuse stuff (that is more convenient than 
4. Make overloads for `autoDiffToGradientMatrix`, etc., that accept symbolic vectors.

For usage in `pydrake`, 4 may be the most straightforward solution for now, as it should be relatively easy to implement in `pydrake`.

I believe Tobias mentioned Cassadi and Python's [ad](https://pypi.org/project/ad/) package which may do 3.a, and he has found convenient.
@TobiaMarcucci When you have a chance, could you post some links to examples of those usages here? (to track for solutions like 3.a or 3.b)

\cc @avalenzu

**Comment by TobiaMarcucci** [comment]
Yes, my goal as said is to linearize the dynamics of a `RigidBodyTree` around a point that is not an equilibrium, and I'm working in python.

At the moment I'm doing a huge workaround to get the linearization, since from python I'm not able to get the inverse of the mass matrix in auto diff form. So, first I linearize the equations in the "manipulator form" then I put the system in state-space form (doing this I can invert the mass matrix only at the nominal configuration).

I agree that, for this specific case, the best solution would be to have a relaxed `Linearize`. Otherwise: is there somewhere a function that returns `x_dot` given `(q, v, u, f_ext)`? (I didn't find one.) In that case I could just use the `jacobian` function in [here](https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/forwarddiff.py) I guess.

For reference this is [CasADi](https://github.com/casadi/casadi/wiki). I've been using it a couple of years ago, and it's a rather user-friendly AD and optimal control package. The thing I liked is that on the high level it seems to work with in symbolic (in the sense that I didn't have to track the indices of the variables I was using), however, I don't know if this poses some problems on the performance side?
